lib/Parser/BOP/multiply.pm

```diff
@@ -68,6 +68,21 @@ sub _reduce {
 	$self->swapOps
 		if (($self->{rop}->class eq 'Number' && $self->{lop}->class ne 'Number' && $reduce->{'x*n'})
 			|| ($self->{lop}->class eq 'Function' && $self->{rop}->class ne 'Function' && $reduce->{'fn*x'}));
+
+	if ($reduce->{'m*(n*x)'}
+		&& defined $self->{lop}
+		&& defined $self->{rop}{lop}
+		&& defined $self->{rop}{bop}
+		&& $self->{lop}->class eq 'Number'
+		&& $self->{rop}{lop}->class eq 'Number'
+		&& $self->{rop}{bop} eq '*')
+	{
+		my $m = $self->{lop};
+		$self->{lop} = $self->{rop}->copy;
+		$self->{lop}->swapOps;
+		$self->{lop}{lop} = $m;
+		$self->{rop} = $self->{rop}{rop};
+	}
```

Comment:
This is a good idea, and one that certainly will help derivatives simplify.  It is one of several that I had been meaning to add, but never got around to it.

There are couple of issues, here, though.  First, the condition for the if-then is not as clean as it should be.  First, you don't need to test that `$self->{lop}` is defined, as that is always the case (the `new` method always defines `lop`, and if it isn't defined, something serious has gone wrong.

Then you should use `$self->{rop}->isa('Parser::BOP::multiply')` to test if the right operator is a multiplication rather than testing if it has a `bop` equal to `*`.

Finally, manipulating the structure of the current object isn't advised.  Rather, the process used by other reduction rules it so create a new sub-tree with the needed structure and return that.  That makes sure that all the proper flags and such are properly set (e.g., the `isConstant` marker, its type, and so on).  For example, in this case, the `rop` is likely not a constant, so its copy would not be marked constant, but when you adjust the copy's operands, they are now both constant, and `$self->{lop}{isConstant}` should be true, but won't be.  If you are in a context that has overridden some of the classes, then other other values may also be incorrectly set.  

So I'd suggest changing this to something like

``` perl
        if ($reduce->{'m*(n*x)'}
            && $self->{lop}->class eq 'Number'
            && $self->{rop}->isa('Parser::BOP::multiply')
            && $self->{rop}{lop}->class eq 'Number'
           ) {
          $self = $self->Item('BOP')->new(
             $self->{equation}, self->{bop},
             $self->Item('BOP')->new($self->{equation}, $self->{bop}, $self->{lop}, $self->{rop}{lop}),
             $self->{rop}{rop}
          )->reduce;
        }
```

rather than manipulating the existing tree.

Reply:
Thanks Davide! One question about all this (this morning after coffee); does it make sense to do it all right here, versus create a method for it? Like there is the `swapOps` method. I started thinking there could be an `associateOps` method and it could also be applied to `m+(n+x)`. You could pass it something indicating if you want to regroup moving grouping to the left or to the right. If I created a method like that, does it belong in the same place as `swapOps` is defined?

Reply:
Along with `m*(n*x)` and `m+(n+x)`, there's also a need for `(x+m)+n`. We already have `x*n` and I think that removes any need for something like `(x*m)*n`. Is there anything else along these lines to add? How do we feel about `m*(x/n)` becoming `m/n*x`? Or `1/x*y` becoming `y/x`?

Reply:
> does it make sense to do it all right here, versus create a method for it?

If you end up needing this code in more than one place, then yes, a method would be useful, and it can go in BOP.pm like `swapOps`.

> it could also be applied to `m+(n+x)`

Yes, but I suspect you might want a rule that does `n+x => x+n` (where x is not a number), and then you can just looks for `(x+n)+m` rather than all the the permutations (`m+(x+n)`, `m+(n+x)`, `(n+x)+m`).  You could also do `(x-n)+m => (x+(m-n)`, `(x+n)-m => x+(n-m)`, and `(x-n)-m => x-(n+m)`.  

I'm OK with the fraction reductions.  Would you also want `m*(n/x)=>(m*n)/x` and similar changes?  There are lots of possibilities.  You just have to be sure your don't have rules that undo other rules; e.g., negatives are factored out so that they can be combined via `-(-x)=>x` or rules like `x-(-y)=>x+y`, so you don't want a rule that puts negatives back in, as that could cause an infinite loop as the rules cycle the result.

I think you can removes the `->reduce` in my code above, as the only thing that can be reduced is the product `m*n` and that will be taken care of in the new multiplication BOP automatically.

Reply:
`n+x => x+n` makes me uneasy because people's polynomials like `1+x+x^2` that they intentionally arranged in that order will become `x+1+x^2`. They could turn off the reduction rule, but now they'd have to edit their existing problems. Or this rule could be off by default, but that would be a new precedent (a reduction rule that is off by default). I'm not sure what to think.

Reply:
So would people prefer `n+x => x+n` or `x+n => n+x`? Without one of these, there will still be lots of unsimplified things. I could add *both* `m+(n+x)` and `(x+m)+n`, and without one of `n+x => x+n` or `x+n => n+x`, something like `Formula("1+((2+x)+3)+4")` will still not reduce at all.

Maybe it should be set up so that both `n+x => x+n` and `x+n => n+x` are available, but of course both should not be used at the same time. But I'm not sure the best way to do that. The reduction rule could be called `n+x` but then there could be a context flag that if set, makes `n+x` behave like `x+n`.

After this issue is sorted out, I'll think about the combinations involving a `*` and `/`; and those involving `+` and `-` (subtraction).